server/controllers/EventController.js

```diff
@@ -16,6 +15,64 @@ const getEvents = async (req, res, next) => {
 
   return res.json(events)
 }
+/*
+* Returns an array of JSON objects of a given length (10 if undefined), filled firs by all events that still need submitting, with the rest
+* of the array being filled with the most recent events ordered by date.
+* */
+const getEventDashboardList = async (req, res, next) => {
+
+  // Check if an array length number has been defined and store that value, otherwise default the value to 10
+  let listLength = (req.query.listLength !== undefined) ? req.query.listLength : 10
+
+  // Find & store all events from the collection
+  let allEvents = await getEventsFromDb()
+
+  // Find & store all events that still need submitting
+  let noRecordEvents = allEvents.filter((event) => {
+    return !event.hasAttendanceRecords
+  })
+
+  // Find & store all events that have allready been submitted
+  let recordEvents = allEvents.filter((event) => {
+    return event.hasAttendanceRecords
+  })
+
+  let sortedRecordEvents = recordEvents.sort((a, b) => {
+    const dateA = new Date(a.date)
+    const dateB = new Date(b.date)
+    return (dateA > dateB) ? -1 : 1
+  })
+
+  // Find how many recent events will fit into the final array, and reduce the array to only contain those objects
+  if (noRecordEvents.length >= listLength) {
+    recordEvents = []
+  } else if (recordEvents.length > listLength - noRecordEvents.length) {
+    recordEvents = recordEvents.slice(0, listLength - noRecordEvents.length)
+  }
+
+  let currentEvent;
+
+  let combinedEvents = [...recordEvents, ...noRecordEvents]
+
+  if (req.query.eventId !== null) {
+
+
+    // Using (==) operator here, as even._id and req.query.eventId are different types, and that is not important for the comparison.
+    const doesExist = combinedEvents.some(event => event._id == req.query.eventId)
+    if (!doesExist) {
+      currentEvent = allEvents.filter(event => event._id == req.query.eventId)
+      combinedEvents = [...combinedEvents, ...currentEvent]
```

Comment:
Let's replace this with:

```
const eventToAdd = allEvents.find(event => event._id == req.query.eventId)
combinedEvents.push(eventToAdd)
```

This is because:

- `currentEvent` is going to be a single event because eventId's are unique, so performing a filter operation is going to be more intensive than a single find operation, as find will return the first instance and then end the loop.
- What `combinedEvents = [...combinedEvents, ...currentEvent...]` does is it actually makes a copy of combinedEvents, and currentEvents, and then pushes each of these into the new combinedEvents array (making a copy by value, and not by reference). Since we're only adding one extra item, `.push(currentEvent)` would probably be more descriptive, as well as _slightly_ more performant

Reply:
@Ray-F We should talk about this.